File: 111. Minimum Depth of Binary Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int minDepth(TreeNode root) {
        if(root == null) return 0;
        Queue<TreeNode> q = new LinkedList<>();
        q.offer(root);
        int level = 1;
        int size = 1;
        while(!q.isEmpty()){
            TreeNode top = q.poll();
            if(top.left == null && top.right == null) return level;
            
            if(top.left != null) q.offer(top.left);
            if(top.right != null) q.offer(top.right);
            size--;
            if(size == 0){
                level++;
                size = q.size();
            }
        }

        return level;
    }
}

File: 111. Minimum Depth of Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int minDepth(TreeNode* root) {
        if (root == nullptr) return 0;

        queue<TreeNode*> q;
        q.push(root);
        int level = 1;
        int size = 1;

        while (!q.empty()) {
            TreeNode* top = q.front();
            q.pop();
            if (top->left == nullptr && top->right == nullptr) return level;

            if (top->left) q.push(top->left);
            if (top->right) q.push(top->right);

            size--;
            if (size == 0) {
                level++;
                size = q.size();
            }
        }

        return level;
    }
};
